# ADM - HW3
## What is the best anime in the world?

In this homework we had to gather information about animes and build a search engine to find the matching ones given a query.


#### Libraries

First let's import some libraries and our custom modules

In [1]:
import pandas as pd
import numpy as np
import csv
import nltk
from functions import *

### 1.Data Collection

#### 1.1 Get the list of animes

We gathered our data on the website [MyAnimeList](https://myanimelist.net/topanime.php), a page which collects information about almost all the anime ever produced. First of all we run a function which goes through every webpage of the Top Rated list and saves the urls in a txt file called **links.txt**.

In [2]:
create_txt()

In [3]:
with open('links.txt', 'r', encoding = 'utf-8') as lf:
    num_lines = sum(1 for line in lf)
print(num_lines)

19119


#### 1.2 Crawl Animes

With our script we gathered **19119 links**.
Next up we downloaded the html page for every link and saved them in the **html** folder of the repository, dividing them in batches of 50 htmls and naming them **article_i.html** where *i* is the index of the anime . To do this we wrote a function that uses the *Request* module. This operation took some time and after about 1500 links the website closed connection with us so we had to go there and manually solve the CAPTCHA. To avoid the script downloading the CAPTCHA page we added a check that warns us at which file stopped so that we could continue from there. 


In [4]:
download_pages()

#### 1.3 Parse downloaded Pages

After we downloaded all the html pages we wrote a function that gathers the information we were interested. These are all the informations we took:
1. Anime Name (saved as animeTitle): String
2. Anime Type (saved as animeType): String
3. Number of episode (saved as animeNumEpisode): Integer
4. Release and End Dates of anime (saved as releaseDate and endDate): Convert both release and end date into datetime format.
5. Number of members (saved as animeNumMembers): Integer
6. Score (saved as animeScore): Float
7. Users (saved as animeUsers): Integer
8. Rank (saved as animeRank): Integer
9. Popularity (saved as animePopularity): Integer
10. Synopsis (saved as animeDescriptions): String
11. Related Anime (saved as animeRelated): List of strings.
12. Characters (saved as animeCharacters): List of strings.
13. Voices (saved as animeVoices): List of strings
14. Staff (saved as animeStaff): Include the staff name and their responsibility/task in a list of lists.
15. Url page (saved as Url): String





and stores them in a tsv file called **anime_i.tsv** where *i* is the index of the anime. These files were saved in the **tsv** folder in batches of 50 tsv for subfolder.





In [5]:
parse_pages()


To work easily with all the data we merged all the tsv in a single **merged.tsv** file using a custom function. 

In [6]:
merging()


In [7]:
df = pd.read_csv('merged.tsv', delimiter = '\t')
df

Unnamed: 0                                         animeTitle  \
0               0                   Fullmetal Alchemist: Brotherhood   
1               1                                 Gintama: The Final   
2               2                                           Gintama.   
3               3                         3-gatsu no Lion 2nd Season   
4               4                                     Koe no Katachi   
...           ...                                                ...   
19114       19114                                   Konbini Shoujo Z   
19115       19115                                   Korogashi Ryouta   
19116       19116                           Kyonyuu Elf Oyako Saimin   
19117       19117                 Mahou Shoujo Elena DVD-BOX Special   
19118       19118  Mama x Holic: Miwaku no Mama to Amaama Kankei ...   

      animeType animeNumEpisode releaseDate     endDate  animeNumMembers  \
0            TV              64  2009-04-05  2010-07-04        2672149.0   
1         Movie               1  2021-01-08  Unnamed: 4          79027.0   
2            TV              12  2017-01-09  2017-03-27         245830.0   
3            TV              22  2017-10-14  2018-03-31         323800.0   
4         Movie               1  2016-09-17  Unnamed: 4        1775911.0   
...         ...             ...         ...         ...              ...   
19114       OVA      Unnamed: 2  Unnamed: 3  Unnamed: 4            554.0   
19115       OVA               3  1990-11-21  1991-11-21            309.0   
19116       OVA               2  2022-01-28  Unnamed: 4            100.0   
19117   Special               1  2015-12-11  Unnamed: 4            327.0   
19118       OVA               2  2021-11-26  Unnamed: 4           1000.0   

       animeScore  animeUsers   animeRank  animePopularity  \
0            9.16     1622384           1              3.0   
1             9.0       29979          10           1924.0   
2            8.98      108581          11            725.0   
3            8.97      155163          12            528.0   
4            8.97     1208990          13             23.0   
...           ...         ...         ...              ...   
19114  Unnamed: 6  Unnamed: 7  Unnamed: 8          12775.0   
19115  Unnamed: 6  Unnamed: 7  Unnamed: 8          14509.0   
19116  Unnamed: 6  Unnamed: 7  Unnamed: 8          17769.0   
19117  Unnamed: 6  Unnamed: 7  Unnamed: 8          14499.0   
19118  Unnamed: 6  Unnamed: 7  Unnamed: 8          11115.0   

                                       animeDescriptions  \
0      After a horrific alchemy experiment goes wrong...   
1                                     New Gintama movie.   
2      After joining the resistance against the bakuf...   
3      Now in his second year of high school, Rei Kir...   
4      As a wild youth, elementary school student Sho...   
...                                                  ...   
19114                                        Unnamed: 10   
19115  Takao Ryota is a pervy but well-meaning biker-...   
19116                                        Unnamed: 10   
19117  Short special included in the DVD-BOX of Mahou...   
19118                                        Unnamed: 10   

                                            animeRelated  \
0      ['Fullmetal Alchemist', 'Fullmetal Alchemist: ...   
1      ['Gintama', 'Gintama.: Shirogane no Tamashii-h...   
2        ['Gintama', 'Gintama°', 'Gintama.: Porori-hen']   
3             ['3-gatsu no Lion', 'Lion meets HachiClo']   
4          ['Koe no Katachi', 'Koe no Katachi Specials']   
...                                                  ...   
19114                                        Unnamed: 11   
19115                               ['Korogashi Ryouta']   
19116                                        Unnamed: 11   
19117                             ['Mahou Shoujo Elena']   
19118  ['Love x Holic: Miwaku no Otome to Hakudaku Ka...   

                                         a

### 2. Search Engine
#### 2.1 Conjunctive Query

After we got our tsv file with informations about all of the animes we started building the Search Engine. First of all we created a file called **vocabulary.tsv** which maps every word in every entry of the animeDescriptions column to a unique ID.
To do this we first pre-processed all the informations in the tsv file by removing stopwords, punctuation and stemming. We also created a file called **dictionary.json** which, for every ID in the **vocabulary.tsv** file, has a list of all the animes that contains the word corresponding to the ID. They are created the first time a User run the **query** function if the files do not exist yet.

In [8]:
query(input('Insert query : '))

Insert query : saiyan race


Index                                         animeTitle  \
364     365                                      Dragon Ball Z   
399     400                           Dragon Ball Super: Broly   
1467   1468  Dragon Ball Z Special 1: Tatta Hitori no Saish...   

                                      animeDescriptions  \
364   Five years after winning the World Martial Art...   
399   Forty-one years ago on Planet Vegeta, home of ...   
1467  Bardock, Son Goku's father, is a low-ranking S...   

                                                    Url  
364     https://myanimelist.net/anime/813/Dragon_Ball_Z  
399   https://myanimelist.net/anime/36946/Dragon_Bal...  
1467  https://myanimelist.net/anime/986/Dragon_Ball_...

#### 2.2 Conjunctive query & Ranking score

Next up we upgraded the Search Engine by implementing a similarity score of the document given a query. To do this we created two other files, **term_idf.json** which contains the idf score for every word in the **vocabulary.tsv** and the **inverted_index.json** which assigns for every word in the vocabulary a list of tuples, each of them consisting of two values, the document ID and the idfscore of that word for that anime. With these new files we calculated the cosine similarity between all the files found by using the custom **query** function and our new query.


In [9]:
similarity(input('Insert query : '))

Insert query : story experiment


Index                         animeTitle  \
129     130                           Nichijou   
1695   1696  Tanoshii Muumin Ikka Bouken Nikki   
1584   1585       Saenai Heroine no Sodatekata   
5         6               Clannad: After Story   
1479   1480                   Majo no Tabitabi   

                                      animeDescriptions  \
129   Nichijou primarily focuses on the daily antics...   
1695  In the remote and mysterious Moomin Valley liv...   
1584  Tomoya Aki, an otaku, has been obsessed with c...   
5     Clannad: After Story, the sequel to the critic...   
1479  Since childhood, Elaina has always been fascin...   

                                                    Url  similarity  
129        https://myanimelist.net/anime/10165/Nichijou    1.000000  
1695  https://myanimelist.net/anime/5712/Tanoshii_Mu...    1.000000  
1584  https://myanimelist.net/anime/23277/Saenai_Her...    0.966805  
5     https://myanimelist.net/anime/4181/Clannad__Af...    0.943407  
1479  https://myanimelist.net/anime/40571/Majo_no_Ta...    0.943407

### Define a new score

We built a new metric to define a similarity score between a given query and the documents we have. This time we took into account all the information contained in the tsv file and not only the *animeDescription*. The way we defined the score was comparing the query with each value of the tsv and add a certain quantity if there was a matching. These are the parameters we chose:

 * title -> score += (1/len(t))*2
 * animetype ->  score += (1/len(ty))*1.5
 * episode -> score += 0.5
 * released -> score += (1/len(d1))
 * endD -> score += (1/len(d2))
 * score -> score += 0.5
 * users -> score += 0.5
 * rank -> score += 0.5
 * popularity -> score += 0.5
 * descriptions -> already scored
 * related -> score += 1/len(r)
 * characters -> score += (1/len(c))*1.5
 * voices ->  score += (1/len(v))
 * staff -> score += (1/len(a))*2
 
 

In [10]:
search(input('Insert query : '))

Insert query : sport run


Index                           animeTitle  \
159      160                           Cross Game   
2147    2148               Taishou Yakyuu Musume.   
16793  16794  Kkoma Bus Tayo-ui Ace Guchuljakjeon   
11730  11731                          Try Knights   
4132    4133        Prince of Stride: Alternative   

                                       animeDescriptions  \
159    Kou Kitamura and Aoba Tsukishima are often at ...   
2147   In 1925 (year 14 of the Taisho era) baseball i...   
16793  It all starts when Duri lost his favorite spor...   
11730  Riku Haruma enters high school without a futur...   
4132   "Stride"—an extreme sport that combines parkou...   

                                                     Url     score  
159        https://myanimelist.net/anime/5941/Cross_Game  1.000000  
2147   https://myanimelist.net/anime/5141/Taishou_Yak...  1.000000  
16793                                                NaN  1.000000  
11730    https://myanimelist.net/anime/39527/Try_Knights  1.000000  
4132   https://myanimelist.net/anime/31559/Prince_of_...  0.958559

### Algorithmic Question

You consult for a personal trainer who has a back-to-back sequence of requests for appointments. A sequence of requests is of the form > 30, 40, 25, 50, 30, 20 where each number is the time that the person who makes the appointment wants to spend. You need to accept some requests, however you need a break between them, so you cannot accept two consecutive requests. For example, [30, 50, 20] is an acceptable solution (of duration 100), but [30, 40, 50, 20] is not, because 30 and 40 are two consecutive appointments. Your goal is to provide to the personal trainer a schedule that maximizes the total length of the accepted appointments. For example, in the previous instance, the optimal solution is [40, 50, 20], of total duration 110.

In [11]:
def AlgorithmicQuestion(Input):
    #initzialization of variables
    Inclusive = 0 #this will keep track of the sum INCLUDING the element of the given Index
    Exclusive = 0 #this will keep track of the sum EXCLUDING the element of the given Index
    Difference = [] #the array of the difference between Inclsive and Exclusive; this quantity is key in finding the elements that maximize the sum of non adjacent elements
    Path = [] #the array of non adjacent elements that maximize the sum 
    
    #computation to maximizes the total length of the accepted appointments
    for Index in Input:
        Temp = Inclusive #dummy variable
        Inclusive = max(Inclusive, Exclusive + Index) #the sum of the elements that maximize the sum; only the maximun value is recorded to be compared in the next iteration
        Exclusive = Temp #aka the previous maximun value
        Difference.append(Inclusive - Exclusive)
    
    #computation of the accepted appointments
    for Index in reversed(range(len(Input))): #we analyze the vector from the end because only at the end we know which appointments were selected
        if Difference[Index] > 0:
            Path.append(Input[Index])
            Difference[Index - 1] = 0 #this is necessary to not consider the adjacent element
    Path.reverse()
    
    #output
    print(f"The Appointments are: {Input}\n\
The personal trainer needs to chose the appointments: {Path}\n\
This will result in a total amount of time of: {max(Inclusive, Exclusive)}")

In [12]:
AlgorithmicQuestion([30, 40, 25, 50, 30, 20])

The Appointments are: [30, 40, 25, 50, 30, 20]
The personal trainer needs to chose the appointments: [40, 50, 20]
This will result in a total amount of time of: 110


In [13]:
from random import randint
n = 10
AlgorithmicQuestion([randint(0,50) for i in range(n)])


The Appointments are: [33, 10, 40, 45, 26, 17, 41, 49, 42, 6]
The personal trainer needs to chose the appointments: [33, 40, 26, 41, 42]
This will result in a total amount of time of: 182
